source/boolean-algebra/sec-boolean-functions.ptx

```diff
@@ -1,26 +1,179 @@
+        <input>
+            h_idempotent.truthtable()
+        </input>
+    </sage>
+    <sage>
+        <input>
+            # Why is this False?
+            h_idempotent.truthtable() == h.truthtable()
+        </input>
+    </sage>
+    <p>
+        Now let's examine the truth table of the Boolean function <m>h(x, y) = x + \bar{y}</m>.
+    </p>
+    <sage>
+        <input>
+            h = formula("x | ~y")
+            h.truthtable()
+        </input>
+    </sage>
+    <p>
+        Using boolean identities is one way of finding the sum-of-products expansion of a Boolean function. Let's find a more programmatic way to find the sum-of-products expansion of the Boolean function. The <c>truthtable()</c> function is not iterable, so we must represent the table as a <c>list</c> with <c>get_table_list()</c>.
+    </p>
+    <sage>
+        <input>
+            from sage.logic.propcalc import formula
+            h = formula("x | ~y")
+            truth_table = h.truthtable()
+            table_list = truth_table.get_table_list()
+        </input>
+    </sage>
+    <p>
+        Now, we can iterate over the truth table to find the Sum-of-Products expansion of the Boolean function. For each row where the output is <c>True</c>, we will construct a product term by examining the input values. For each variable, if its value is <c>True</c>, we include the variable in the product; if its value is <c>False</c>, we include its negation in the product.
+     </p>
+     <sage>
+        <input>
+            # Initialize the SOP expansion
+            sop_expansion = []
+
+            # Iterate through the truth table
+            for row in table_list[1:]:  # Skip the header row ['x', 'y']
+                if row[-1]:  # If the output is True
+                    # Construct the minterm
+                    minterm = []
+                    for var, value in zip(table_list[0], row[:-1]):  # Iterate over variables and their values
+                        if value:
+                            minterm.append(var)  # Include variable as is if True
+                        else:
+                            minterm.append(f'~{var}')  # Include negation if False
+                    
+                    sop_expansion.append(' &amp; '.join(minterm))  # Combine using AND
+        </input>
+     </sage>
+     <p>
+        Now, we can combine all minterms:
+     </p>
+     <sage>
+        <input>
+            sop_result = ' | '.join(f'({m})' for m in sop_expansion)
+            print("SOP expansion:", sop_result)
+        </input>
+     </sage>
+     <p>
+        Let's verify that the sum-of-products expansion we found with the truth table is the same as the one we found using boolean identities.
+     </p>
+     <sage>
+        <input>
+            sop_result == h
+        </input>
+        <output>
+            
+        </output>
+     </sage>
     <idx><h>polynomial</h></idx>
     <sage>
         <input>
             B = BooleanPolynomialRing(3, 'x')

```

Comment:
Is BooleanPolynomialRing something we want to keep?

Reply:
I dont think it is relevant anymore, but let's keep it for now and see what the professor says